### PR TITLE
feat(customer_v2):  add customer create v2 endpoint 

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [x] Refactoring

## Description

Added customer create v2 end point till domain model

## How did you test it?

Test Sanity for v1 customer end-points

- Create
- Retrieve
- Update
- Delete
- Ephemeral Key

Tested customer Create api :

```shell
curl --location 'http://localhost:8080/customers' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_VnlEN2ZNoInmxctmOS18TG3WQCUfbjDycerly6JJxXSp36OxXzrlzVkjYIezVoZT' \
--data-raw '{
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "First customer",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "address" : {
        "city":"siliguri"
    }
}'
```
DB Entry
Screenshot 2024-08-01 at 8 33 48 PM
Screenshot 2024-08-01 at 8 37 00 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

File: crates/hyperswitch_domain_models/src/customer.rs [hunk 1]
```diff
@@ -36,12 +33,47 @@ pub struct Customer {
...
+pub struct Customer {
+    pub merchant_id: id_type::MerchantId,
+    pub name: crypto::OptionalEncryptableName,
+    pub email: crypto::OptionalEncryptableEmail,
+    pub phone: crypto::OptionalEncryptablePhone,
+    pub phone_country_code: Option<String>,
+    pub description: Option<String>,
+    pub created_at: PrimitiveDateTime,
+    pub metadata: Option<pii::SecretSerdeValue>,
+    pub connector_customer: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/customer.rs:50]:
```suggestion
    pub connector_customer: Option<pii::SecretSerdeValue>,
```

File: crates/hyperswitch_domain_models/src/customer.rs [hunk 2]
```diff
@@ -127,10 +161,174 @@ impl super::behaviour::Conversion for Customer {
...
+        name: crypto::OptionalEncryptableName,
+        email: crypto::OptionalEncryptableEmail,
+        phone: Box<crypto::OptionalEncryptablePhone>,
+        description: Option<String>,
+        phone_country_code: Option<String>,
+        metadata: Option<pii::SecretSerdeValue>,
+        connector_customer: Option<serde_json::Value>,
+    },
+    ConnectorCustomer {
+        connector_customer: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/customer.rs:285]:
```suggestion
        connector_customer: Option<pii::SecretSerdeValue>,
```

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/storage_impl/src/lib.rs [hunk 3]
```diff
@@ -440,6 +441,23 @@ impl UniqueConstraints for diesel_models::Customer {
     }
 }
 
+#[cfg(all(feature = "v2", feature = "customer_v2"))]
+impl UniqueConstraints for diesel_models::Customer {
+    fn unique_constraints(&self) -> Vec<String> {
+        vec![format!(
+            "customer_{}_{}",
+            self.merchant_reference_id
```
Comment (Narayanbhat166) [inline crates/storage_impl/src/lib.rs:449]: This should be id

File: crates/storage_impl/src/payouts/payouts.rs [hunk 4]
```diff
@@ -518,6 +531,7 @@ impl<T: DatabaseStore> PayoutsInterface for crate::RouterStore<T> {
     }
 
     #[cfg(feature = "olap")]
+    #[cfg(all(any(feature = "v1", feature = "v2"), not(feature = "customer_v2")))]
```
Comment (Narayanbhat166) [inline crates/storage_impl/src/payouts/payouts.rs:534]: include the olap feature also here

File: crates/router/src/core/payments/customers.rs [hunk 5]
```diff
@@ -72,23 +74,29 @@ pub async fn create_connector_customer<F: Clone, T: Clone>(
...
-        .and_then(|connector_customer_value| connector_customer_value.get(connector_name))
-        .and_then(|connector_customer| connector_customer.as_str())
+        .and_then(|connector_customer_value| {
+            connector_customer_value
+                .clone()
+                .expose()
+                .get(connector_name)
+                .cloned()
+        })
+        .and_then(|connector_customer| serde_json::from_value::<String>(connector_customer).ok())
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/customers.rs:91]: we should not ignore the error here

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

File: crates/router/src/core/customers.rs [hunk 6]
```diff
@@ -238,15 +254,14 @@ impl CustomerCreateBridge for customers::CustomerRequest {
...
             modified_at: common_utils::date_time::now(),
             default_payment_method_id: None,
             updated_by: None,
-            // default_billing_address: default_customer_billing_address,
-            // default_shipping_address: default_customer_shipping_address,
-            // merchant_reference_id,
+            default_billing_address: encrypted_customer_billing_address.map(Into::into),
+            default_shipping_address: encrypted_customer_shipping_address.map(Into::into),
             // status: Some(customer_domain::SoftDeleteStatus::Active)
+            version: common_enums::ApiVersion::V2,
```
Comment (Narayanbhat166) [inline crates/router/src/core/customers.rs:264]: we can have this as a const value, based on the feature flag we can have the const value which will be either v1 or v2

Comment (sahkal) [reply crates/router/src/core/customers.rs:264]: sure

File: crates/router/src/consts.rs [hunk 7]
```diff
@@ -130,3 +130,9 @@ pub const CONNECTOR_CREDS_TOKEN_TTL: i64 = 900;
 
 //max_amount allowed is 999999999 in minor units
 pub const MAX_ALLOWED_AMOUNT: i64 = 999999999;
+
+#[cfg(all(any(feature = "v1", feature = "v2"), not(feature = "customer_v2")))]
```
Comment (Narayanbhat166) [inline crates/router/src/consts.rs:134]: the feature gates will be only v1 and v2

**Review by jarnura — DISMISSED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by Aprabhat19 — DISMISSED** [review]
Payment Method Changes LGTM

File: v2_migrations/2024-07-30-100323_customer_v2/down.sql [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+ALTER TABLE customers DROP COLUMN IF EXISTS merchant_reference_id;
+ALTER TABLE customers DROP COLUMN IF EXISTS default_billing_address;
+ALTER TABLE customers DROP COLUMN IF EXISTS default_shipping_address;
+
+-- Run this query only when V1 is deprecated
+ALTER TABLE customers DROP CONSTRAINT customers_pkey;
+ALTER TABLE customers DROP COLUMN IF EXISTS id;
+ALTER TABLE customers ADD COLUMN IF NOT EXISTS id SERIAL;
+
+ALTER TABLE customers ADD COLUMN customer_id VARCHAR(64) NOT NULL;
```
Comment (SanchithHegde) [inline v2_migrations/2024-07-30-100323_customer_v2/down.sql:11]: This won't work if we have existing data in the table?

**Review by SanchithHegde — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by lsampras — APPROVED** [review]
Analytic events look good

**Pull request merged**